File: swift-bridge/Sources/CoreDataBridge/ModelMetadata.swift
import CoreData
import Foundation

@_cdecl("cd_property_description_new")
public func cdPropertyDescriptionNew(
    _ outProperty: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outProperty?.pointee = cdRetain(NSPropertyDescription())
    return CDR_OK
}

@_cdecl("cd_property_description_get_name")
public func cdPropertyDescriptionGetName(_ propertyPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let propertyPtr else {
        return nil
    }
    let property: NSPropertyDescription = cdBorrow(propertyPtr)
    return cdCString(property.name)
}

@_cdecl("cd_property_description_set_name")
public func cdPropertyDescriptionSetName(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ name: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr, let name else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing property description or name")
        }
        let property: NSPropertyDescription = cdBorrow(propertyPtr)
        property.name = String(cString: name)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_property_description_get_optional")
public func cdPropertyDescriptionGetOptional(_ propertyPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let propertyPtr else {
        return 0
    }
    let property: NSPropertyDescription = cdBorrow(propertyPtr)
    return property.isOptional ? 1 : 0
}

@_cdecl("cd_property_description_set_optional")
public func cdPropertyDescriptionSetOptional(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ optional: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing property description")
        }
        let property: NSPropertyDescription = cdBorrow(propertyPtr)
        property.isOptional = optional != 0
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_property_description_get_transient")
public func cdPropertyDescriptionGetTransient(_ propertyPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let propertyPtr else {
        return 0
    }
    let property: NSPropertyDescription = cdBorrow(propertyPtr)
    return property.isTransient ? 1 : 0
}

@_cdecl("cd_property_description_set_transient")
public func cdPropertyDescriptionSetTransient(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ transient: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing property description")
        }
        let property: NSPropertyDescription = cdBorrow(propertyPtr)
        property.isTransient = transient != 0
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_property_description_new")
public func cdFetchedPropertyDescriptionNew(
    _ outProperty: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outProperty?.pointee = cdRetain(NSFetchedPropertyDescription())
    return CDR_OK
}

@_cdecl("cd_fetched_property_description_get_fetch_request")
public func cdFetchedPropertyDescriptionGetFetchRequest(_ propertyPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let propertyPtr else {
        return nil
    }
    let property: NSFetchedPropertyDescription = cdBorrow(propertyPtr)
    return property.fetchRequest.map(cdRetain)
}

@_cdecl("cd_fetched_property_description_set_fetch_request")
public func cdFetchedPropertyDescriptionSetFetchRequest(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ fetchRequestPtr: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched property description")
        }
        let property: NSFetchedPropertyDescription = cdBorrow(propertyPtr)
        property.fetchRequest = fetchRequestPtr.map { cdBorrow($0) as NSFetchRequest<NSFetchRequestResult> }
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_expression_description_new")
public func cdExpressionDescriptionNew(
    _ outProperty: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outProperty?.pointee = cdRetain(NSExpressionDescription())
    return CDR_OK
}

@_cdecl("cd_expression_description_get_result_type")
public func cdExpressionDescriptionGetResultType(_ propertyPtr: UnsafeMutableRawPointer?) -> UInt64 {
    guard let propertyPtr else {
        return 0
    }
    let property: NSExpressionDescription = cdBorrow(propertyPtr)
    return UInt64(property.expressionResultType.rawValue)
}

@_cdecl("cd_expression_description_set_result_type")
public func cdExpressionDescriptionSetResultType(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ resultType: UInt64,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing expression description")
        }
        let property: NSExpressionDescription = cdBorrow(propertyPtr)
        guard let resultType = NSAttributeType(rawValue: UInt(resultType)) else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Invalid expression result type")
        }
        property.expressionResultType = resultType
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_derived_attribute_description_new")
public func cdDerivedAttributeDescriptionNew(
    _ outAttribute: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outAttribute?.pointee = cdRetain(NSDerivedAttributeDescription())
    return CDR_OK
}

@_cdecl("cd_composite_attribute_description_new")
public func cdCompositeAttributeDescriptionNew(
    _ outAttribute: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    guard #available(macOS 14.0, *) else {
        cdWriteError(cdUnavailableFeatureError("NSCompositeAttributeDescription"), to: outError)
        return CDR_FAILURE
    }
    outAttribute?.pointee = cdRetain(NSCompositeAttributeDescription())
    return CDR_OK
}

@_cdecl("cd_composite_attribute_description_get_elements")
public func cdCompositeAttributeDescriptionGetElements(_ attributePtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard #available(macOS 14.0, *), let attributePtr else {
        return nil
    }
    let attribute: NSCompositeAttributeDescription = cdBorrow(attributePtr)
    return cdRetain(attribute.elements as NSArray)
}

@_cdecl("cd_composite_attribute_description_set_elements")
public func cdCompositeAttributeDescriptionSetElements(
    _ attributePtr: UnsafeMutableRawPointer?,
    _ elementPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard #available(macOS 14.0, *) else {
            throw cdUnavailableFeatureError("NSCompositeAttributeDescription")
        }
        guard let attributePtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing composite attribute description")
        }
        let attribute: NSCompositeAttributeDescription = cdBorrow(attributePtr)
        let elements: [NSAttributeDescription] = cdObjects(from: elementPtrs, count: count)
        attribute.elements = elements
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_element_description_new")
public func cdFetchIndexElementDescriptionNew(
    _ propertyPtr: UnsafeMutableRawPointer?,
    _ collationType: UInt64,
    _ outElement: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let propertyPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index element property")
        }
        guard let collationType = NSFetchIndexElementType(rawValue: UInt(collationType)) else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Invalid fetch index element collation type")
        }
        let property: NSPropertyDescription = cdBorrow(propertyPtr)
        outElement?.pointee = cdRetain(NSFetchIndexElementDescription(property: property, collationType: collationType))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_element_description_get_property_name")
public func cdFetchIndexElementDescriptionGetPropertyName(_ elementPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let elementPtr else {
        return nil
    }
    let element: NSFetchIndexElementDescription = cdBorrow(elementPtr)
    return element.propertyName.flatMap(cdCString)
}

@_cdecl("cd_fetch_index_element_description_get_collation_type")
public func cdFetchIndexElementDescriptionGetCollationType(_ elementPtr: UnsafeMutableRawPointer?) -> UInt64 {
    guard let elementPtr else {
        return 0
    }
    let element: NSFetchIndexElementDescription = cdBorrow(elementPtr)
    return UInt64(element.collationType.rawValue)
}

@_cdecl("cd_fetch_index_element_description_set_collation_type")
public func cdFetchIndexElementDescriptionSetCollationType(
    _ elementPtr: UnsafeMutableRawPointer?,
    _ collationType: UInt64,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let elementPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index element description")
        }
        guard let collationType = NSFetchIndexElementType(rawValue: UInt(collationType)) else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Invalid fetch index element collation type")
        }
        let element: NSFetchIndexElementDescription = cdBorrow(elementPtr)
        element.collationType = collationType
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_element_description_get_ascending")
public func cdFetchIndexElementDescriptionGetAscending(_ elementPtr: UnsafeMutableRawPointer?) -> Int32 {
    guard let elementPtr else {
        return 0
    }
    let element: NSFetchIndexElementDescription = cdBorrow(elementPtr)
    return element.isAscending ? 1 : 0
}

@_cdecl("cd_fetch_index_element_description_set_ascending")
public func cdFetchIndexElementDescriptionSetAscending(
    _ elementPtr: UnsafeMutableRawPointer?,
    _ ascending: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let elementPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index element description")
        }
        let element: NSFetchIndexElementDescription = cdBorrow(elementPtr)
        element.isAscending = ascending != 0
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_description_new")
public func cdFetchIndexDescriptionNew(
    _ name: UnsafePointer<CChar>?,
    _ elementPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int32,
    _ outIndex: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let name else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index description name")
        }
        let elements: [NSFetchIndexElementDescription] = cdObjects(from: elementPtrs, count: count)
        outIndex?.pointee = cdRetain(NSFetchIndexDescription(name: String(cString: name), elements: elements))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_description_get_name")
public func cdFetchIndexDescriptionGetName(_ indexPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let indexPtr else {
        return nil
    }
    let index: NSFetchIndexDescription = cdBorrow(indexPtr)
    return cdCString(index.name)
}

@_cdecl("cd_fetch_index_description_set_name")
public func cdFetchIndexDescriptionSetName(
    _ indexPtr: UnsafeMutableRawPointer?,
    _ name: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let indexPtr, let name else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index description or name")
        }
        let index: NSFetchIndexDescription = cdBorrow(indexPtr)
        index.name = String(cString: name)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetch_index_description_get_elements")
public func cdFetchIndexDescriptionGetElements(_ indexPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let indexPtr else {
        return nil
    }
    let index: NSFetchIndexDescription = cdBorrow(indexPtr)
    return cdRetain(index.elements as NSArray)
}

@_cdecl("cd_fetch_index_description_set_elements")
public func cdFetchIndexDescriptionSetElements(
    _ indexPtr: UnsafeMutableRawPointer?,
    _ elementPtrs: UnsafePointer<UnsafeMutableRawPointer?>?,
    _ count: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let indexPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetch index description")
        }
        let index: NSFetchIndexDescription = cdBorrow(indexPtr)
        let elements: [NSFetchIndexElementDescription] = cdObjects(from: elementPtrs, count: count)
        index.elements = elements
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

File: src/model_metadata.rs
use core::ffi::c_void;

use crate::error::CoreDataError;
use crate::ffi;
use crate::private::{
    collect_array, cstring_from_str, error_from_status, impl_object_wrapper, take_string,
};
use crate::query::NSFetchRequest;
use crate::schema::{AttributeType, NSAttributeDescription};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum NSFetchIndexElementType {
    Binary,
    RTree,
    Unknown(u64),
}

impl NSFetchIndexElementType {
    const fn from_raw(raw: u64) -> Self {
        match raw {
            0 => Self::Binary,
            1 => Self::RTree,
            other => Self::Unknown(other),
        }
    }

    const fn as_raw(self) -> u64 {
        match self {
            Self::Binary => 0,
            Self::RTree => 1,
            Self::Unknown(raw) => raw,
        }
    }
}

impl_object_wrapper!(NSPropertyDescription);
impl_object_wrapper!(NSFetchedPropertyDescription);
impl_object_wrapper!(NSExpressionDescription);
impl_object_wrapper!(NSDerivedAttributeDescription);
impl_object_wrapper!(NSCompositeAttributeDescription);
impl_object_wrapper!(NSFetchIndexDescription);
impl_object_wrapper!(NSFetchIndexElementDescription);

fn clone_retained_wrapper<T>(ptr: *mut c_void, context: &str) -> Result<T, CoreDataError>
where
    T: crate::private::FromRetainedPtr,
{
    let retained = unsafe { ffi::cd_retain_object(ptr) };
    unsafe { T::from_retained_ptr(retained, context) }
}

impl NSPropertyDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_property = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe { ffi::cd_property_description_new(&mut out_property, &mut out_error) };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_property, "property description") }
    }

    pub fn name(&self) -> Result<String, CoreDataError> {
        let ptr = unsafe { ffi::cd_property_description_get_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
            .ok_or_else(|| CoreDataError::bridge(-1, "property description name was nil"))
    }

    pub fn set_name(&self, name: &str) -> Result<(), CoreDataError> {
        let name = cstring_from_str(name, "property description name")?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_property_description_set_name(self.as_ptr(), name.as_ptr(), &mut out_error)
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn is_optional(&self) -> bool {
        unsafe { ffi::cd_property_description_get_optional(self.as_ptr()) != 0 }
    }

    pub fn set_optional(&self, optional: bool) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_property_description_set_optional(
                self.as_ptr(),
                i32::from(optional),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn is_transient(&self) -> bool {
        unsafe { ffi::cd_property_description_get_transient(self.as_ptr()) != 0 }
    }

    pub fn set_transient(&self, transient: bool) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_property_description_set_transient(
                self.as_ptr(),
                i32::from(transient),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }
}

impl NSFetchedPropertyDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_property = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status =
            unsafe { ffi::cd_fetched_property_description_new(&mut out_property, &mut out_error) };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_property, "fetched property description") }
    }

    pub fn fetch_request(&self) -> Result<Option<NSFetchRequest>, CoreDataError> {
        let ptr = unsafe { ffi::cd_fetched_property_description_get_fetch_request(self.as_ptr()) };
        if ptr.is_null() {
            return Ok(None);
        }
        Ok(Some(unsafe {
            NSFetchRequest::from_retained_ptr(ptr, "fetched property fetch request")?
        }))
    }

    pub fn set_fetch_request(
        &self,
        fetch_request: Option<&NSFetchRequest>,
    ) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetched_property_description_set_fetch_request(
                self.as_ptr(),
                fetch_request.map_or(core::ptr::null_mut(), NSFetchRequest::as_ptr),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn as_property_description(&self) -> Result<NSPropertyDescription, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "property description")
    }
}

impl NSExpressionDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_property = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status =
            unsafe { ffi::cd_expression_description_new(&mut out_property, &mut out_error) };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_property, "expression description") }
    }

    pub fn expression_result_type(&self) -> AttributeType {
        AttributeType::from_raw(unsafe {
            ffi::cd_expression_description_get_result_type(self.as_ptr())
        })
    }

    pub fn set_expression_result_type(
        &self,
        result_type: AttributeType,
    ) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_expression_description_set_result_type(
                self.as_ptr(),
                result_type.as_raw(),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn as_property_description(&self) -> Result<NSPropertyDescription, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "property description")
    }
}

impl NSDerivedAttributeDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_attribute = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_derived_attribute_description_new(&mut out_attribute, &mut out_error)
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_attribute, "derived attribute description") }
    }

    pub fn as_attribute_description(&self) -> Result<NSAttributeDescription, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "attribute description")
    }
}

impl NSCompositeAttributeDescription {
    pub fn new() -> Result<Self, CoreDataError> {
        let mut out_attribute = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_composite_attribute_description_new(&mut out_attribute, &mut out_error)
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_attribute, "composite attribute description") }
    }

    pub fn elements(&self) -> Result<Vec<NSAttributeDescription>, CoreDataError> {
        let array_ptr =
            unsafe { ffi::cd_composite_attribute_description_get_elements(self.as_ptr()) };
        collect_array(array_ptr, "composite attribute elements")
    }

    pub fn set_elements(&self, elements: &[&NSAttributeDescription]) -> Result<(), CoreDataError> {
        let raw_elements = elements
            .iter()
            .map(|element| element.as_ptr())
            .collect::<Vec<_>>();
        let count = i32::try_from(raw_elements.len())
            .map_err(|_| CoreDataError::bridge(-1, "composite attribute element count overflow"))?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_composite_attribute_description_set_elements(
                self.as_ptr(),
                raw_elements.as_ptr(),
                count,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn as_attribute_description(&self) -> Result<NSAttributeDescription, CoreDataError> {
        clone_retained_wrapper(self.as_ptr(), "attribute description")
    }
}

impl NSFetchIndexElementDescription {
    pub fn new(
        property: &NSPropertyDescription,
        collation_type: NSFetchIndexElementType,
    ) -> Result<Self, CoreDataError> {
        let mut out_element = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_element_description_new(
                property.as_ptr(),
                collation_type.as_raw(),
                &mut out_element,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_element, "fetch index element description") }
    }

    pub fn property_name(&self) -> Option<String> {
        let ptr =
            unsafe { ffi::cd_fetch_index_element_description_get_property_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
    }

    pub fn collation_type(&self) -> NSFetchIndexElementType {
        NSFetchIndexElementType::from_raw(unsafe {
            ffi::cd_fetch_index_element_description_get_collation_type(self.as_ptr())
        })
    }

    pub fn set_collation_type(
        &self,
        collation_type: NSFetchIndexElementType,
    ) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_element_description_set_collation_type(
                self.as_ptr(),
                collation_type.as_raw(),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn is_ascending(&self) -> bool {
        unsafe { ffi::cd_fetch_index_element_description_get_ascending(self.as_ptr()) != 0 }
    }

    pub fn set_ascending(&self, ascending: bool) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_element_description_set_ascending(
                self.as_ptr(),
                i32::from(ascending),
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }
}

impl NSFetchIndexDescription {
    pub fn new(
        name: &str,
        elements: &[&NSFetchIndexElementDescription],
    ) -> Result<Self, CoreDataError> {
        let name = cstring_from_str(name, "fetch index description name")?;
        let raw_elements = elements
            .iter()
            .map(|element| element.as_ptr())
            .collect::<Vec<_>>();
        let count = i32::try_from(raw_elements.len())
            .map_err(|_| CoreDataError::bridge(-1, "fetch index element count overflow"))?;
        let mut out_index = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_description_new(
                name.as_ptr(),
                raw_elements.as_ptr(),
                count,
                &mut out_index,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_index, "fetch index description") }
    }

    pub fn name(&self) -> Result<String, CoreDataError> {
        let ptr = unsafe { ffi::cd_fetch_index_description_get_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
            .ok_or_else(|| CoreDataError::bridge(-1, "fetch index description name was nil"))
    }

    pub fn set_name(&self, name: &str) -> Result<(), CoreDataError> {
        let name = cstring_from_str(name, "fetch index description name")?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_description_set_name(self.as_ptr(), name.as_ptr(), &mut out_error)
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn elements(&self) -> Result<Vec<NSFetchIndexElementDescription>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_fetch_index_description_get_elements(self.as_ptr()) };
        collect_array(array_ptr, "fetch index description elements")
    }

    pub fn set_elements(
        &self,
        elements: &[&NSFetchIndexElementDescription],
    ) -> Result<(), CoreDataError> {
        let raw_elements = elements
            .iter()
            .map(|element| element.as_ptr())
            .collect::<Vec<_>>();
        let count = i32::try_from(raw_elements.len())
            .map_err(|_| CoreDataError::bridge(-1, "fetch index element count overflow"))?;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetch_index_description_set_elements(
                self.as_ptr(),
                raw_elements.as_ptr(),
                count,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }
}
